### PR TITLE
Changelog test

### PR DESCRIPTION
**Pull request by SamMousa**
Because I never get the changelog right on the first go, a unit test that evaluates each line.

I noticed some (older) lines do not obey the changelog; rather than I forcing all old lines to be reformatted I have added an offset.

| Q             | A
| ------------- | ---
| Is bugfix?    | no
| New feature?  | no
| Breaks BC?    | no
| Tests pass?   | no

### CODE REVIEW

**Comment by MysteryDragon** [comment]
If this test could also checks right order of strings... It will be wonderful as for me :)

**Comment by SamMousa** [comment]
What requirement is there on the order?

File: tests/framework/ChangeLogTest.php [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+{
+    public function changeProvider()
+    {
+
+        $lines = explode("\n", file_get_contents(__DIR__ . '/../../framework/CHANGELOG.md'));
+
+        // Don't check last 1500 lines, they are old and often don't obey the standard.
+        $lastIndex = count($lines) - 1500;
+        $result = [];
+        foreach($lines as $i => $line) {
```
Comment (MysteryDragon) [inline tests/framework/ChangeLogTest.php]: @SamMousa let me be menace and say that it's better to put space after `foreach` for PSR-2 compability :)

**Comment by MysteryDragon** [comment]
> What requirement is there on the order?
I just noticed that maintainers often reorders CHANGELOG:
https://github.com/yiisoft/yii2/commit/d44d5540b1f58b92ac1f9c6ac1ffc0036970026a
(for example)
It looks like it's Bug -> Enh -> Chg order, and in each group numbers sorts by ascendent.

**Comment by SamMousa** [comment]
Now we also test the order of lines, note that an error is only thrown on the first erroneous line (since we just compare the current line type vs the previous one).

**Comment by samdark** [comment]
Lines are automatically reordered on release. It doesn't make sense to require correct order from contributors at the moment PR is created.

**Comment by SamMousa** [comment]
Ah okay, so i'll remove that then; any other feedback?

**Comment by MysteryDragon** [comment]
@samdark oh, it's sad for me. I like strict rules =/
@SamMousa sorry for confusing you

**Comment by samdark** [comment]
https://github.com/yiisoft/yii2/blob/master/build/controllers/ReleaseController.php

**Comment by SamMousa** [comment]
@samdark I think it's ready now; reverted the last 2 commits.

**Comment by rob006** [comment]
> Lines are automatically reordered on release. It doesn't make sense to require correct order from contributors at the moment PR is created.

Changelog is reordered multiple times between releases, and each reordering creates conflicts with existing PRs. It would save a lot of pain if changelog were always in the right order.

It would be great if @yii-bot could reorder changelog in PR...

**Comment by SamMousa** [comment]
> Changelog is reordered multiple times between releases, and each reordering creates conflicts with existing PRs. It would save a lot of pain if changelog were always in the right order.
>
>It would be great if @yii-bot could reorder changelog in PR...

Alternatively it would be great if the reordering was done ONLY just before a release.

**Comment by MysteryDragon** [comment]
> Alternatively it would be great if the reordering was done ONLY just before a release.
I guess some conflicts will still exists (PRs which have been created before release, and still open after that). But it's better option too, of course.

**Comment by samdark** [comment]
> Alternatively it would be great if the reordering was done ONLY just before a release.

**Comment by SamMousa** [comment]
Ah, i didn't know. Anyway I've removed the ordering test, it should be okay now!

**Review by samdark — APPROVED** [review]

**Comment by samdark** [comment]
Let's have it. Merged. Thank you.

**Pull request merged**